### ISSUE TITLE
Metaclass `__get__` under `TYPE_CHECKING` makes class unusable as base

### ISSUE THREAD

**Issue by jvdillon** [astral-sh/ty#3279]
When a metaclass defines `__get__` under `TYPE_CHECKING`, ty no longer recognizes classes using that metaclass as valid base classes.

Minimal reproducer:

```python
from __future__ import annotations
from typing import TYPE_CHECKING

class MyMeta(type):
    if TYPE_CHECKING:
        def __get__(cls, obj: object, owner: type) -> type:
            return cls

class Base(metaclass=MyMeta):
    pass

class Parent:
    class Config(Base):
        x: int = 0

class Child(Parent):
    class Config(Parent.Config):  # unsupported-base
        y: int = 1
```

```
warning[unsupported-base]: Unsupported class base
  |
  |     class Config(Parent.Config):
  |                  ^^^^^^^^^^^^^ Has type `type`
  |
info: Only class objects or `Any` are supported as class bases
```

Removing the `__get__` method makes the warning disappear. The `__get__` return type doesn't matter — `type`, `type[Any]`, etc. all reproduce it.

**Expected behavior:** `Parent.Config` should still be usable as a base class. The `__get__` descriptor provides type-narrowing for attribute access, but shouldn't change the fact that `Config` is a
class.

**Use case:** This pattern is used to narrow the return type of a factory method on nested config classes (metaclass `__get__` returns an intersection type so the type checker knows what `.make()`
returns). Config inheritance is a core part of the API.

**ty version:** 0.0.30

Note: This feature is needed for my project, [configgle](https://github.com/rekursiv-ai/configgle).

**Comment by AlexWaygood** [comment]
Thanks for the report! It looks to me like the issue here is that the return type of your metaclass `__get__` method is annotated incorrectly? If `__get__` always returns `cls`, you can change `__get__` to be generic, like this:

```py
from __future__ import annotations
from typing import TYPE_CHECKING, TypeVar

T = TypeVar("T")

class MyMeta(type):
    if TYPE_CHECKING:
        def __get__(cls: T, obj: object, owner: type) -> T:
            return cls

class Base(metaclass=MyMeta):
    pass

class Parent:
    class Config(Base):
        x: int = 0

class Child(Parent):
    class Config(Parent.Config):  # unsupported-base
        y: int = 1
```

And the error [goes away](https://play.ty.dev/8e4d753b-0f31-4edb-bbf4-1d9eced1c0c3).

The issue here is that `Base` (and all subclasses of `Base`) are instances of `MyMeta`, so therefore `Base` (and all subclasses of `Base`) are descriptor instances. ty therefore needs to use the descriptor protocol to infer the type that the attribute access `Parent.Config` will resolve to -- i.e., it cannot simply assume that `Parent.Config` will be the class it was defined as; it must lookup the return type of `Parent.Config.__get__`. This is why:

```pycon
>>> class MyMeta(type):
...     def __get__(cls, owner, instance):
...         return 42
...         
>>> class Outer:
...     class Inner(metaclass=MyMeta): ...
...     
>>> Outer.Inner
42
>>> 
```

**Comment by jvdillon** [comment]
You're right! Tysm for the help--this has been an issue for a while in fact!
